stout/http.h

```diff
@@ -0,0 +1,529 @@
+
+struct Result {
+  std::string data;
+  long response;
+};
+
+////////////////////////////////////////////////////////////////////////
+
+namespace detail {
+
+////////////////////////////////////////////////////////////////////////
+
+enum class Method {
+  GET,
+  POST,
+};
+
+////////////////////////////////////////////////////////////////////////
+
+class EasyHandle {
+ public:
+  EasyHandle() {
+    CHECK_NOTNULL(easy_ = curl_easy_init());
+  }
+
+  EasyHandle(const EasyHandle& that) = delete;
+  EasyHandle(EasyHandle&& that) {
+    easy_ = that.easy_;
+    that.easy_ = nullptr;
+  }
+
+  ~EasyHandle() {
+    if (easy_ == nullptr) {
+      return;
+    }
+
+    curl_easy_cleanup(easy_);
+  }
+
+  CURL* handle() {
+    CHECK_NOTNULL(easy_);
+    return easy_;
+  }
+
+ private:
+  CURL* easy_ = nullptr;
+
+  struct WriteCallbackData {
+    WriteCallbackData() = default;
+
+    WriteCallbackData(const WriteCallbackData& that) = delete;
+
+    WriteCallbackData(WriteCallbackData&& that) {
+      buffer = std::move(that.buffer);
+      should_interrupt.store(
+          that.should_interrupt.load(std::memory_order_relaxed),
+          std::memory_order_relaxed);
+    }
+
+    EventLoop::Buffer buffer;
+    std::atomic<bool> should_interrupt = false;
+  };
+
+  // Called whenever easy handle gets incoming data.
+  // Return value should always be (size * nmemb) unless you have to interrupt
+  // this particular connection, in which case (nmemb * size - 1) is used
+  // as a return value.
+  static size_t WriteCallback(
+      char* ptr, // Pointer to incoming data. Memory is handled by CURL.
+      size_t size,
+      size_t nmemb,
+      WriteCallbackData* userdata // curl_easy_setopt WRITEDATA
+  ) {
+    EventLoop::Buffer& buffer = userdata->buffer;
+
+    buffer += std::string(ptr, size * nmemb);
+
+    return userdata->should_interrupt.load(std::memory_order_relaxed)
+        ? nmemb * size - 1
+        : nmemb * size;
+  }
+
+  friend auto HttpCall(
+      EventLoop&,
+      const std::string&,
+      const Method&,
+      const std::chrono::nanoseconds&,
+      const std::string&);
+};
+
+////////////////////////////////////////////////////////////////////////
+
+class MultiHandle {
+ public:
+  MultiHandle() {
+    CHECK_NOTNULL(multi_ = curl_multi_init());
+  }
+
+  MultiHandle(const MultiHandle& that) = delete;
+  MultiHandle(MultiHandle&& that) {
+    multi_ = that.multi_;
+    that.multi_ = nullptr;
+  }
+
+  ~MultiHandle() {
+    if (multi_ == nullptr) {
+      return;
+    }
+
+    CHECK_EQ(curl_multi_cleanup(multi_), CURLM_OK);
+  }
+
+  CURLM* handle() {
+    CHECK_NOTNULL(multi_);
+    return multi_;
+  }
+
+ private:
+  CURLM* multi_ = nullptr;
+};
+
+////////////////////////////////////////////////////////////////////////
+
+inline auto HttpCall(
+    EventLoop& loop,
+    const std::string& url,
+    const Method& method,
+    const std::chrono::nanoseconds& timeout = std::chrono::nanoseconds(0),
```

Comment:
To simplify this code let's handle timers outside. We can create an abstraction like:
```
auto e = DoUnlessTimeout(
    http::Get(...),
    std::chrono::milliseconds(500));
```
Which can then be used by multiple folks that want to be able to pass in an eventual that can be interrupted.